### PR TITLE
feat: add validations for 1.1 authorization models

### PR DESCRIPTION
**Pull request by craigpastro**
## Description

Needs https://github.com/openfga/api/pull/27 to be merged first.

## Review Checklist
- [ ] I have clicked on ["allow edits by maintainers"](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork).
- [ ] I have added documentation for new/changed functionality in this PR or in a PR to [openfga.dev](https://github.com/openfga/openfga.dev) [Provide a link to any relevant PRs in the references section above]
- [ ] The correct base branch is being used, if not `main`
- [ ] I have added tests to validate that the change in functionality is working as expected

### CODE REVIEW

File: pkg/cmd/service/service_test.go [hunk 1]
```diff
@@ -590,8 +590,6 @@ func TestHTTPServingTLS(t *testing.T) {
 			KeyPath:  certsAndKeys.serverKeyFile,
 		}
 
-		fmt.Println(config)
```
Comment (elbuo8) [inline pkg/cmd/service/service_test.go:593]: ?

Comment (craigpastro) [reply pkg/cmd/service/service_test.go:593]: I think I added this in on some PR way back when and now I am trying to sneak the deletion of it in this PR.

File: server/commands/write_authzmodel.go [hunk 2]
```diff
@@ -29,14 +31,23 @@ func NewWriteAuthorizationModelCommand(
...
+
 	typeDefinitions := req.GetTypeDefinitions().GetTypeDefinitions()
 
 	// Until this is solved: https://github.com/envoyproxy/protoc-gen-validate/issues/74
 	if len(typeDefinitions) > w.backend.MaxTypesInTypeDefinition() {
 		return nil, serverErrors.ExceededEntityLimit("type definitions in an authorization model", w.backend.MaxTypesInTypeDefinition())
 	}
 
-	if err := validateAuthorizationModel(typeDefinitions); err != nil {
+	if err := validateAuthorizationModel(req.GetSchemaVersion(), typeDefinitions); err != nil {
```
Comment (elbuo8) [inline server/commands/write_authzmodel.go:50]: `schemaVersion` is already defined above.

Comment (craigpastro) [reply server/commands/write_authzmodel.go]: Oops. Thank you! Fixed.

File: server/commands/write_authzmodel.go [hunk 3]
```diff
@@ -163,3 +185,83 @@ func isUsersetRewriteValid(allRelations map[string]struct{}, relationsOnType map
...
+	for _, td := range tds {
+		for _relation, _ := range td.GetRelations() {
+			metadata, ok := td.GetMetadata().GetRelations()[_relation]
+			if ok {
+				for _, userType := range metadata.GetDirectlyRelatedUserTypes() {
+					objectType := userType.GetType()
+					relation := userType.GetRelation()
+
+					if _, ok := typeToRelations[objectType]; !ok {
+						return serverErrors.InvalidRelationType(td.GetType(), _relation, objectType, relation)
```
Comment (miparnisari) [inline server/commands/write_authzmodel.go:211]: i think these two errors should be different, because at this point, `relation` can be empty

File: server/test/write_authzmodel.go [hunk 4]
```diff
@@ -337,28 +470,78 @@ func TestWriteAuthorizationModel(t *testing.T, datastore storage.OpenFGADatastor
...
+}
+
+func TestWriteAuthorizationModelWithRelationalTypes(t *testing.T, datastore storage.OpenFGADatastore) {
+	var tests = []struct {
+		name    string
+		request *openfgapb.WriteAuthorizationModelRequest
+		err     error
+	}{
+		{
+			name: "succeeds",
```
Comment (miparnisari) [inline server/test/write_authzmodel.go:526]:
```suggestion
			name: "succeeds when DirectlyRelatedUserTypes has only a type",
```

File: server/commands/write_authzmodel.go (hunk 3 above)
Comment (craigpastro) [reply server/commands/write_authzmodel.go:211]: It is the type of the relation that is invalid so I think this makes sense. Suppose the type is `user` and then:

1. Without relation the error is "the type "user" is invalid on ..."
2. With a relation, say `member`, the error is "the type "user#member` is invalid on ..."

WDYT?

File: server/test/write_authzmodel.go [hunk 5]
```diff
@@ -367,37 +550,141 @@ func TestWriteAuthorizationModel(t *testing.T, datastore storage.OpenFGADatastor
 					},
 				},
 			},
-			err: errors.RelationNotFound("owner", "", nil),
 		},
 		{
-			_name: "ExecuteWriteFailsIfTupleToUsersetReferencesUnknownRelation",
+			name: "succeeds part 2",
```
Comment (miparnisari) [inline server/test/write_authzmodel.go:555]:
```suggestion
			name: "succeeds when DirectlyRelatedUserTypes has a type and a relation",
```

File: server/test/write_authzmodel.go [hunk 6]
```diff
@@ -494,38 +796,73 @@ func TestWriteAuthorizationModel(t *testing.T, datastore storage.OpenFGADatastor
...
+			err: errors.InvalidRelationType("document", "reader", "user", ""),
 		},
 	}
+
+	ctx := context.Background()
+	logger := logger.NewNoopLogger()
+
 	for _, test := range tests {
-		t.Run(test._name, func(t *testing.T) {
+		t.Run(test.name, func(t *testing.T) {
```
Comment (miparnisari) [inline server/test/write_authzmodel.go:865]: please send a different PR with this change 😬

File: server/commands/write_authzmodel.go (hunk 3 above)
Comment (miparnisari) [reply server/commands/write_authzmodel.go:211]: You mean that you would use `InvalidRelationType` in both but format the error nicely? If so, 👍

File: server/commands/write_authzmodel.go [hunk 7]
```diff
@@ -163,3 +185,83 @@ func isUsersetRewriteValid(allRelations map[string]struct{}, relationsOnType map
...
+			typeToRelations[typeName][relationName] = struct{}{}
+		}
+	}
+
+	// Here we are checking that every type on a relation is valid. This means:
+	// 1. If it is a type (e.g. user) then this type is a type in the model
+	// 2. If it is a type#relation then this type with this relation is in the model
+	for _, td := range tds {
+		for _relation, _ := range td.GetRelations() {
+			metadata, ok := td.GetMetadata().GetRelations()[_relation]
```
Comment (miparnisari) [inline server/commands/write_authzmodel.go:204]: if people give us some relations with types defined but others with nothing (i.e. `td.GetMetadata() == nil`), we should throw an error

File: server/test/write_authzmodel.go [hunk 8]
```diff
@@ -367,37 +550,141 @@ func TestWriteAuthorizationModel(t *testing.T, datastore storage.OpenFGADatastor
...
+				StoreId:       storeID,
+				TypeDefinitions: []*openfgapb.TypeDefinition{
+					{
+						Type: "document",
+						Relations: map[string]*openfgapb.Userset{
+							"reader": {Userset: &openfgapb.Userset_This{}},
+						},
+						Metadata: &openfgapb.Metadata{
+							Relations: map[string]*openfgapb.RelationalMetadata{
 								"reader": {
```
Comment (miparnisari) [inline server/test/write_authzmodel.go:650]: can you add a test for the case where the auth model has 2 types but only one has defined metadata? (we should not allow it)

Comment (miparnisari) [inline server/test/write_authzmodel.go:650]: what happens if the relation `owner` exists but is not defined in the `metadata` map?

File: server/commands/write_authzmodel.go (hunk 7 above)
Comment (craigpastro) [reply server/commands/write_authzmodel.go:204]: Yes. In this case, if so, then `ok` should be false.

File: server/commands/write_authzmodel.go [hunk 9]
```diff
@@ -60,7 +65,13 @@ func (w *WriteAuthorizationModelCommand) Execute(ctx context.Context, req *openf
...
 //     a. Be valid relations on the same type in the authorization model (in cases of computedUserset)
 //     b. Be valid relations on another existing type (in cases of tupleToUserset)
-func validateAuthorizationModel(tds []*openfgapb.TypeDefinition) error {
+//
+// If it is a 1.1 model (with types on relations), additionally:
+//  3. Every type on a relation must be a valid type:
+//     a. For a type (e.g. user) this means checking that this type is in the model
+//     b. For a type#relation this means checking that this type with this relation is in the model
+//  4. Check that a relation is assignable if and only if it has a non-zero list of types
+func validateAuthorizationModel(schemaVersion typesystem.SchemaVersion, tds []*openfgapb.TypeDefinition) error {
```
Comment (jon-whit) [inline server/commands/write_authzmodel.go:74]: Let's put this and similar methods under the `typesystem` package since it is logic specific to the validation of types in an authorization model.

Comment (craigpastro) [reply server/commands/write_authzmodel.go:74]: I love this idea, but do you mind if I do it in a following PR? I have one PR to follow this which will introduce the `typesystem` package and get the schema version into the datastore and that has already blown up like crazy.

File: server/test/write_authzmodel.go (hunk 4 above)
Comment (craigpastro) [reply server/test/write_authzmodel.go:547]: Updated. Thanks!

File: server/test/write_authzmodel.go (hunk 5 above)
Comment (craigpastro) [reply server/test/write_authzmodel.go:576]: Updated. Thanks!

File: server/test/write_authzmodel.go (hunk 6 above)
Comment (craigpastro) [reply server/test/write_authzmodel.go:828]: https://github.com/openfga/openfga/pull/238

File: server/test/check.go [hunk 10]
```diff
@@ -1400,7 +1400,7 @@ func TestCheckQuery(t *testing.T, datastore storage.OpenFGADatastore) {
 	data, err := os.ReadFile(gitHubTestDataFile)
 	require.NoError(t, err)
 
-	var gitHubTypeDefinitions openfgapb.TypeDefinitions
+	var gitHubTypeDefinitions openfgapb.WriteAuthorizationModelRequest
```
Comment (jon-whit) [inline server/test/check.go:1403]: The json data contained in `./testdata/github.json` encodes one or more type definitions, but it doesn't encode a `WriteAuthorizationModelRequest`.

This should remain as is should it not?

File: server/commands/write_authzmodel.go (hunk 9 above)
Comment (jon-whit) [reply server/commands/write_authzmodel.go:74]: I'd prefer we do it here since it is technically part of the scope of this work. This PR introduces the `typesystem` package and since this method is a static method, there shouldn't be any reason why we can't just quickly copy/paste it to that package and then reference it. We're already importing the `typesystem` package in this code, so we may as well move this to where it belongs now. I don't see a need to postpone that for another PR.

File: pkg/typesystem/typesystem.go [hunk 11]
```diff
@@ -0,0 +1,31 @@
...
+
+type SchemaVersion int
+
+const (
+	SchemaVersionUnspecified SchemaVersion = iota
+	SchemaVersion1_0
+	SchemaVersion1_1
+)
+
+func NewSchemaVersion(s string) SchemaVersion {
```
Comment (jon-whit) [inline pkg/typesystem/typesystem.go:11]: Please add a `typesystem_test.go` test file and test this and the `String()` method below.

File: server/test/check.go (hunk 10 above)
Comment (craigpastro) [reply server/test/check.go:1403]: We don't have an `openfgapb.TypeDefinitions` anymore, which is why this changed. I imagine in the future when we update these models to have types we will need to include the schema version as well.

File: server/test/server.go [hunk 12]
```diff
@@ -62,11 +62,8 @@ func RunQueryTests(t *testing.T, ds storage.OpenFGADatastore) {
 
 func RunCommandTests(t *testing.T, ds storage.OpenFGADatastore) {
 	t.Run("TestWriteCommand", func(t *testing.T) { TestWriteCommand(t, ds) })
-
-	t.Run("TestWriteAuthorizationModel",
-		func(t *testing.T) { TestWriteAuthorizationModel(t, ds) },
-	)
-
+	t.Run("TestWriteAuthorizationModel", func(t *testing.T) { TestWriteAuthorizationModel(t, ds) })
+	t.Run("TestWriteAuthorizationModelWithRelationalTypes", func(t *testing.T) { TestWriteAuthorizationModelWithRelationalTypes(t, ds) })
```
Comment (jon-whit) [inline server/test/server.go:66]: Can we refactor this name to `TestWriteAuthorizationModelWithTypeInfo` and the same goes for the actual function name getting called in `server/test/write_authzmodel.go`

File: server/commands/write_authzmodel.go [hunk 13]
```diff
@@ -29,14 +30,18 @@ func NewWriteAuthorizationModelCommand(
 
 // Execute the command using the supplied request.
 func (w *WriteAuthorizationModelCommand) Execute(ctx context.Context, req *openfgapb.WriteAuthorizationModelRequest) (*openfgapb.WriteAuthorizationModelResponse, error) {
-	typeDefinitions := req.GetTypeDefinitions().GetTypeDefinitions()
+	schemaVersion := typesystem.NewSchemaVersion(req.GetSchemaVersion())
+	if schemaVersion == typesystem.SchemaVersionUnspecified {
```
Comment (jon-whit) [inline server/commands/write_authzmodel.go:34]: Existing clients will not be providing a `schema_version` field in their requests, and thus this will default to `SchemaVersionUnspecified`, which will lead to this error every time.

We need to account for this to make sure we don't break people's existing integrations. What are your thoughts?

Comment (jon-whit) [reply server/commands/write_authzmodel.go:34]: Nevermind. We return 1.0 on "". Sorry

File: server/commands/write_authzmodel.go (hunk 9 above)
Comment (jon-whit) [inline server/commands/write_authzmodel.go:74]: Mind changing this to take an `AuthorizationModel` instead of type definitions?

Comment (jon-whit) [reply server/commands/write_authzmodel.go:74]: Also, do you mind changing this to take an `AuthorizationModel` instead of type definitions?

File: server/commands/write_authzmodel.go [hunk 14]
```diff
@@ -163,3 +180,83 @@ func isUsersetRewriteValid(allRelations map[string]struct{}, relationsOnType map
 
 	return nil
 }
+
+func areRelationalTypesValid(tds []*openfgapb.TypeDefinition) error {
+	typeToRelations := map[string]map[string]struct{}{}
```
Comment (jon-whit) [inline server/commands/write_authzmodel.go:185]: There should be a helper method for this in the `typesystem` package.

I think we should have something like

```
package typesystem

type Typesystem struct {
    model *openfgapb.AuthorizationModel
}

func New(model *openfga.AuthorizationModel) *Typesystem {
    ....
}

func (t *Typesystem) GetTypeRelationsMap() map[string]map[string]Relation {
    // logic here
}
```

File: server/commands/write_authzmodel.go (hunk 7 above)
Comment (jon-whit) [reply server/commands/write_authzmodel.go:199]: If `ok` is false this indicates the relation had no type information associated with it, correct?

e.g.

```
type user

type document
  relations
    define reader: [user]
    define can_read: reader
```

Is that correct?

File: server/commands/write_authzmodel.go [hunk 15]
```diff
@@ -163,3 +180,83 @@ func isUsersetRewriteValid(allRelations map[string]struct{}, relationsOnType map
...
+					return serverErrors.NonassignableRelationHasAType(td.GetType(), relation)
+				}
+			}
+		}
+	}
+
+	return nil
+}
+
+func isAssignable(rewrite *openfgapb.Userset) bool {
```
Comment (jon-whit) [inline server/commands/write_authzmodel.go:238]: This should also be a helper in the `typesystem` package.

```
package typesystem

type Typesystem struct {...}

func (t *Typesystem) IsAssignableRelation(objectType, relation string) (bool, error) {
   // logic here
}
```

File: server/commands/write_authzmodel.go [hunk 16]
```diff
@@ -163,3 +180,83 @@ func isUsersetRewriteValid(allRelations map[string]struct{}, relationsOnType map
...
+	switch rw := rewrite.GetUserset().(type) {
+	case *openfgapb.Userset_This:
+		return true
+	case *openfgapb.Userset_Union:
```
Comment (jon-whit) [inline server/commands/write_authzmodel.go:242]: If we used `UsersetRewrite_SetOp` here instead of individual `Union`, `Intersection`, `Difference` this would collapse into

```
case *openfgapb.UsersetRewrite_SetOp:
    for _, child := rw.GetSetOperation().GetChild() {
       if isAssignable(child) { return true }
    }
```
Another +1 opportunity for that refactor at some point 😄

File: server/test/write_authzmodel.go [hunk 17]
```diff
@@ -362,28 +491,78 @@ func TestWriteAuthorizationModel(t *testing.T, datastore storage.OpenFGADatastor
...
+}
+
+func TestWriteAuthorizationModelWithRelationalTypes(t *testing.T, datastore storage.OpenFGADatastore) {
+	var tests = []struct {
+		name    string
+		request *openfgapb.WriteAuthorizationModelRequest
+		err     error
+	}{
+		{
+			name: "succeeds on a valid model with a type type",
```
Comment (jon-whit) [inline server/test/write_authzmodel.go:547]: with a type type?

File: server/test/write_authzmodel.go [hunk 18]
```diff
@@ -362,28 +491,78 @@ func TestWriteAuthorizationModel(t *testing.T, datastore storage.OpenFGADatastor
...
+						Type: "document",
+						Relations: map[string]*openfgapb.Userset{
+							"reader": {Userset: &openfgapb.Userset_This{}},
+						},
+						Metadata: &openfgapb.Metadata{
+							Relations: map[string]*openfgapb.RelationMetadata{
+								"reader": {
+									DirectlyRelatedUserTypes: []*openfgapb.RelationReference{
+										{
+											Type: "user",
```
Comment (jon-whit) [inline server/test/write_authzmodel.go:565]: I've been using help methods for things like this to shrink it from 3 lines to 1.

e.g. `relationref("user", "")`
or `relationref("group", "member")`

File: server/test/write_authzmodel.go [hunk 19]
```diff
@@ -392,37 +571,107 @@ func TestWriteAuthorizationModel(t *testing.T, datastore storage.OpenFGADatastor
...
 										},
 									},
 								},
+								"member": {
+									DirectlyRelatedUserTypes: []*openfgapb.RelationReference{
+										{
+											Type: "user",
```
Comment (jon-whit) [inline server/test/write_authzmodel.go:602]: Sorry, did you mean to have `relationref("group", "member")`. 

The test is called "succeeds on a valid model with a type and type#relation type" but I don't see a `type#relation` reference.

I'm assuming you're testing

```
type user

type group
  relations
    define admin: [user]
    define member: [user, group#member]
```
Is that correct?

Also, as a general note, it would be really cool if we could provide the DSL syntax as a string in these tests for validation 😄 That would be so nice!

@rhamzeh this is another opportunity why I feel it would be good for us to leverage WASM with the DSL transformer. If we could pass in a DSL string in our tests and run it through a DSL transformer wasm module loaded in Go which outputs the protobuf model, then our tests like this could be easier to read and understand.

File: server/commands/write_authzmodel.go (hunk 7 above)
Comment (craigpastro) [reply server/commands/write_authzmodel.go:199]: Correct. Or for some reason the JSON doesn't contain it.

File: server/commands/write_authzmodel.go (hunk 9 above)
Comment (craigpastro) [reply server/commands/write_authzmodel.go:74]: Alright. Let's do it 🚀

File: server/test/write_authzmodel.go (hunk 19 above)
Comment (jon-whit) [reply server/test/write_authzmodel.go:602]: It would become:

```
tests := []struct{
  ...
  model string
}{
  {
    ...
    model: `
        type user

        type document
          relations
            define reader: [user]
    `
  },
}
```
instead of writing out the protobuf structs everywhere.

File: server/test/write_authzmodel.go [hunk 20]
```diff
@@ -7,24 +7,16 @@ import (
...
-func TestWriteAuthorizationModel(t *testing.T, datastore storage.OpenFGADatastore) {
-	type writeAuthorizationModelTestSettings struct {
-		name    string
-		request *openfgapb.WriteAuthorizationModelRequest
-		err     error
-	}
-
-	ctx := context.Background()
-	logger := logger.NewNoopLogger()
+const storeID = "01GCD5SAJVC2Y38MJBNJ9QVZ60"
```
[the remark this answers is not in the document]
Comment (craigpastro) [reply server/test/write_authzmodel.go:17]: Good idea. Thanks. Updated in a new PR and will update here if we get it in!

File: server/test/server.go (hunk 12 above)
Comment (craigpastro) [reply server/test/server.go:66]: Done. Thanks!

File: server/test/write_authzmodel.go (hunk 17 above)
Comment (craigpastro) [reply server/test/write_authzmodel.go:547]: Yeah. haha. I'll update it to read "with a objectType type".

File: server/test/write_authzmodel.go (hunk 19 above)
Comment (craigpastro) [reply server/test/write_authzmodel.go:602]: If you look a bit more down the `document` type relations have `group#member` in them.

File: server/test/write_authzmodel.go [hunk 21]
```diff
@@ -7,24 +7,16 @@ import (
...
 
+func TestWriteAuthorizationModel(t *testing.T, datastore storage.OpenFGADatastore) {
```
Comment (jon-whit) [inline server/test/write_authzmodel.go:19]: Can we add a test that uses TupleToUserset to verify the `tupleset` relation validates against a defined relation as well?

File: server/commands/write_authzmodel.go (hunk 15 above)
Comment (jon-whit) [reply server/commands/write_authzmodel.go:238]: Can we add some more tests for this that make sure behavior under intersection, union, and exclusion are working as expected? I didn't see any rewrites that involved union, intersection, or exclusion in the newly added tests for this functionality.

File: server/test/write_authzmodel.go (hunk 21 above)
Comment (craigpastro) [reply server/test/write_authzmodel.go:19]: Doesn't `succeeds part II` have such?

File: server/commands/write_authzmodel.go (hunk 7 above)
Comment (jon-whit) [reply server/commands/write_authzmodel.go:199]: If it doesn't contain it then we assume it is not assignable, and if it references `self` then the logic below (determining it's assignability) would fail yeah?

File: server/test/write_authzmodel.go (hunk 18 above)
Comment (jon-whit) [reply server/test/write_authzmodel.go:565]: Yeah, probably `typesystem`. I think that's a decent home, because it's referring to relations on a specific object type.

File: server/test/write_authzmodel.go (hunk 19 above)
Comment (jon-whit) [reply server/test/write_authzmodel.go:602]: Yeah, sorry I missed that. That's why I deleted the first part of my comment 😄 . Sorry about that. The comment about the DSL still applies though. That'd be cool to do at a later time.

File: server/test/write_authzmodel.go (hunk 21 above)
Comment (jon-whit) [reply server/test/write_authzmodel.go:19]: Sorry, I put this on the wrong test. I meant for the tests with type restrictions.

I don't see a test that involves union, intersection, exclusion or tuple_to_userset with assignable vs non-assignable types.

Comment (jon-whit) [reply server/test/write_authzmodel.go:19]: Ideally these kind of tests would belong on the helper methods in the `typesystem` package though like we discussed. That way it's more clear what paths of these critical helpers have been fully tested.

File: pkg/typesystem/typesystem.go (hunk 11 above)
Comment (craigpastro) [reply pkg/typesystem/typesystem.go:16]: Tests added. Thanks!

File: server/commands/write_authzmodel.go (hunk 14 above)
Comment (craigpastro) [reply server/commands/write_authzmodel.go]: Thanks! As discussed, let's refactor in a follow up PR.

File: server/commands/write_authzmodel.go (hunk 15 above)
Comment (craigpastro) [reply server/commands/write_authzmodel.go]: Tests added. Please take a look!

File: server/test/write_authzmodel.go (hunk 18 above)
Comment (craigpastro) [reply server/test/write_authzmodel.go:569]: Again, in a follow up PR. Is this `typesystem_testutils` stuff? How do we arrange all this?

File: pkg/typesystem/typesystem.go [hunk 22]
```diff
@@ -0,0 +1,251 @@
...
+		err = isUsersetRewriteValid(allRelations, relationsOnType, objectType, relation, t.Difference.Subtract)
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func areRelationalTypesValid(tds []*openfgapb.TypeDefinition) error {
```
Comment (jon-whit) [inline pkg/typesystem/typesystem.go:173]: Can we please refactor this so that it's easier to understand and read?

I think it makes more sense for this code to be composed like so:

```
func (t *Typesystem) GetRelation(
    objectType, relation string,
) (*openfgapb.Relation, error) {
    // todo: implement me
}

func (t *Typesystem) DirectlyRelatedUserTypes(
    objectType, relation string,
) ([]*openfgapb.RelationReference, error) {
	// todo: implement me
}

func (t *Typesystem) IsDirectlyAssignable(relation *openfgapb.Relation) (bool, error) {
    return containsSelf(relation.GetRewrite())
}

func containsSelf(rewrite *openfgapb.Userset) bool {
    // same code as isAssignable in this PR
}

func (t *Typesystem) ValidateRelationTypeRestrictions(objectType, relation string) error {
    relation, err := t.GetRelation(objectType, relation)
    if err != nil {
        // return an error because objectType or relation is undefined
    }

    relatedTypes := relation.GetTypeInfo().GetDirectlyRelatedUserTypes()

    assignable := t.IsDirectlyAssignable(relation)
    if assignable && len(relatedTypes) < 1 {
        // return an error since there were no type restrictions specified
    }

    if !assignable && len(relatedTypes) != 0 {
        // return an error since there were type restrictions on an unassignable type
    }

    for _, relatedType := range relatedTypes {
        relatedObjType := relatedType.GetObjectType()
        relatedRelation := relatedType.GetRelation()
        
        if relatedRelation != "" {
            _, err := t.GetRelation(relatedObjType, relatedRelation))
            if err != nil {
                // related object type or relation does not exist, return appropriate err
            }

            continue
        }

        _, err := t.GetTypeDefinition(relatedObjType)
        if err != nil {
            // related type does not exist, return appropriate err
        }		
    }

    return nil
}
```

The way it is written now is quite hard to read because you have to cut through so much boilerplate map code to understand the primary important pieces.

**Review by jon-whit — CHANGES_REQUESTED** [review]
Left one last comment about code composition and readability. Let's refactor the code a bit within the scope of this PR so that it leverages more reusable code pieces and makes the code easier to read, and then I think we're good.

Thanks for adding the tests!

File: pkg/typesystem/typesystem.go (hunk 22 above)
Comment (jon-whit) [reply pkg/typesystem/typesystem.go:173]: Related to https://github.com/openfga/api/pull/33 as well. Makes this composition possible.

File: server/test/write_authzmodel.go (hunk 18 above)
Comment (jon-whit) [reply server/test/write_authzmodel.go:580]: I have a follow up PR that will add this. I'll submit that once we merge this 👍

File: server/commands/write_authzmodel.go (hunk 14 above)
Comment (jon-whit) [reply server/commands/write_authzmodel.go]: I left a comment here:
https://github.com/openfga/openfga/pull/223#discussion_r974494567

I don't think this work should get put off, because it's part of the review of this code. I feel we should make the composition of this code more readable before merging, because it makes the code easier to read and maintain and also lays the foundation for further developments to make use of more reusable software pieces related to type definitions.

The last remaining bit that I feel should be done is just a slight refactor like I recommended below. I think once we make it a little more readable through better composition, then we're good to merge 👍

File: pkg/typesystem/typesystem.go [hunk 23]
```diff
@@ -0,0 +1,251 @@
...
+func (m *AuthorizationModel) ToProto() *openfgapb.AuthorizationModel {
+	return &openfgapb.AuthorizationModel{
+		Id:              m.ID,
+		SchemaVersion:   m.Version.String(),
+		TypeDefinitions: m.TypeDefinitions,
+	}
+}
+
+// Validate validates the model according to the following rules:
+//  1. Do not allow duplicate types (or duplication relations but that is inherent in the map structure)
```
[the remark this answers is not in the document]
Comment (craigpastro) [reply pkg/typesystem/typesystem.go:53]: Thank you! Fixed.

File: server/commands/write_authzmodel.go [hunk 24]
```diff
@@ -29,137 +30,37 @@ func NewWriteAuthorizationModelCommand(
...
-		err := isUsersetRewriteValid(allRelations, relationsOnType, objectType, relation, t.Difference.Base)
-		if err != nil {
-			return err
-		}
-
-		err = isUsersetRewriteValid(allRelations, relationsOnType, objectType, relation, t.Difference.Subtract)
-		if err != nil {
-			return err
-		}
+	utils.LogDBStats(ctx, w.logger, "WriteAuthzModel", 0, 1)
```
[the remark this answers is not in the document]
Comment (craigpastro) [reply server/commands/write_authzmodel.go:58]: I don't think I changed the location of this function. I believe that it is where it is since even if the write fails this log will be sent. I am not sure I agree with that, but 🤷

File: server/commands/write_authzmodel.go [hunk 25]
```diff
@@ -29,137 +30,42 @@ func NewWriteAuthorizationModelCommand(
...
+	if err := w.backend.WriteAuthorizationModel(ctx, req.GetStoreId(), id, typeSystem.GetTypeDefinitions()); err != nil {
```
[the remark this answers is not in the document]
Comment (craigpastro) [reply server/commands/write_authzmodel.go:64]: I want to do this in a follow up PR as it is quite a lot and it would be nice to clean up the interfaces as
```
WriteAuthorizationModel(ctx, req.GetStoreId(), id, schemaVersion, typeSystem.GetTypeDefinitions())
```
is getting a bit long. I had such a follow up PR before but it will need a lot of ❤️  now.

**Review by jon-whit — APPROVED** [review]
LGTM for now. We can introduce a follow up PR to handle some minor improvements to the typesystem package and also store the schema version etc..
